packages/react-keyring/src/Authenticator.tsx

```diff
@@ -0,0 +1,132 @@
+   * register a new space or log in using `email`
+   */
+  handleRegisterSubmit?: (e: React.FormEvent<HTMLFormElement>) => Promise<void>
+}
+
+export type AuthenticatorContextActions = KeyringContextActions & {
+  /**
+   * Set an email to be used to log in or register.
+   */
+  setEmail: React.Dispatch<React.SetStateAction<string>>
+}
+
+export type AuthenticatorContextValue = [
+  state: AuthenticatorContextState,
+  actions: AuthenticatorContextActions
+]
+
+export const AuthenticatorContext = createContext<AuthenticatorContextValue>([
+  {
+    spaces: [],
+    submitted: false
+  },
+  {
+    setEmail: () => { throw new Error('missing set email function') },
+    loadAgent: async () => { },
+    unloadAgent: async () => { },
+    resetAgent: async () => { },
+    createSpace: async () => { throw new Error('missing keyring context provider') },
+    setCurrentSpace: async () => { },
+    registerSpace: async () => { },
+    cancelRegisterSpace: () => { },
+    getProofs: async () => []
+  }
+])
+
+/**
+ * Top level component of the headless Authenticator.
+ *
+ * Must be used inside a KeyringProvider.
+ *
+ * Designed to be used by Authenticator.Form, Authenticator.EmailInput
+ * and others to make it easy to implement authentication UI.
+ */
+export function Authenticator (props: any): JSX.Element {
+  const [state, actions] = useKeyring()
+  const { createSpace, registerSpace } = actions
+  const [email, setEmail] = useState('')
+  const [submitted, setSubmitted] = useState(false)
+
+  const handleRegisterSubmit = useCallback(async (e: React.FormEvent<HTMLFormElement>) => {
+    e.preventDefault()
+    setSubmitted(true)
+    try {
+      await createSpace()
+      await registerSpace(email)
+    } catch (err: any) {
+      throw new Error('failed to register')
+    } finally {
+      setSubmitted(false)
+    }
+  }, [setSubmitted, createSpace, registerSpace])
+
+  const value = useMemo<AuthenticatorContextValue>(() => [
+    { ...state, email, submitted, handleRegisterSubmit },
+    { ...actions, setEmail }
+  ], [state, actions, email, submitted, handleRegisterSubmit])
+  return (
+    <AuthenticatorContext.Provider value={value} {...props} />
+  )
+}
+
+/**
+ * Form component for the headless Authenticator.
+ *
+ * A `form` designed to work with `Authenticator`. Any passed props will
+ * be passed along to the `form` component.
+ */
+Authenticator.Form = function Form (props: any) {
+  const [{ handleRegisterSubmit }] = useAuthenticator()
+  return (
+    <form {...props} onSubmit={handleRegisterSubmit} />
+  )
+}
+
+/**
+ * Input component for the headless Uploader.
+ *
+ * An email `input` designed to work with `Authenticator.Form`. Any passed props will
+ * be passed along to the `input` component.
+ */
+Authenticator.EmailInput = function EmailInput (props: any) {
+  const [{ email }, { setEmail }] = useAuthenticator()
+  return (
+    <input type='email' value={email} onChange={e => setEmail(e.target.value)} {...props} />
```

Comment:
Any time we receive arbitrary props we want to spread them first to make sure the user doesn't override anything we don't want them to override. There a few more instances of this.

Reply:
ah yea - my thinking here was that it would be better to give the user the ability to override this stuff if needed, but it's definitely a tradeoff between flexibility and usability - if we opt to let them override we open up more space for accidentally overriding things that are important to the functionality of our component, and if we opt not to we might force the user to abandon/reimplement these components themselves in some cases.

concrete example is the `type` property - we don't technically need that to be `email` for this component to work in a reasonable way, but it is a useful default.

I'm not really sure of the right tradeoff across the board, so I'm inclined to go with your solution and be consistent about always spreading first. If we find users are asking for the ability to override properties, we can change this in the future without breaking backwards compatibility.

Reply:
> I'm inclined to go with your solution and be consistent about always spreading first

That not exactly what I intended to propose above. It's only the things we don't want users to override that should _not_ be overridable. As you correctly identified we might set properties that are safe to be overridden. The concern was more (in this instance) that if we allow `value` or `onChange` to be overridden this component cannot operate correctly.

Reply:
oh got it got it - ok, I agree with that completely! I do still think it's good to start from a place of not letting them override anything until we document what is overridable so that we don't have people depending on behavior we decide to deprecate (ie, going from overridable to not overridable is a breaking change because people might be depending on being able to override something, going from not overridable to overridable isn't) so I'm happy this landed where it did, but I'll feel free to move some props to after the spread in the future!